File: implementations/wrap-rust/src/__tests__/cases/001-module-functions/output/module/serialization.rs
use serde::{Serialize, Deserialize};
use std::convert::TryFrom;
use polywrap_wasm_rs::{
    BigInt,
    BigNumber,
    Map,
    Context,
    DecodeError,
    EncodeError,
    Read,
    ReadDecoder,
    Write,
    WriteEncoder,
    JSON,
};

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ArgsFunction1 {
    pub arg1: u32,
    pub arg2: bool,
}

pub fn deserialize_function1_args(args: &[u8]) -> Result<ArgsFunction1, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing module-type: function1 Args".to_string();

    let mut reader = ReadDecoder::new(args, context);
    let mut num_of_fields = reader.read_map_length()?;

    let mut _arg1: u32 = 0;
    let mut _arg1_set = false;
    let mut _arg2: bool = false;
    let mut _arg2_set = false;

    while num_of_fields > 0 {
        num_of_fields -= 1;
        let field = reader.read_string()?;

        match field.as_str() {
            "arg1" => {
                reader.context().push(&field, "u32", "type found, reading argument");
                _arg1 = reader.read_u32()?;
                _arg1_set = true;
                reader.context().pop();
            }
            "arg2" => {
                reader.context().push(&field, "bool", "type found, reading argument");
                _arg2 = reader.read_bool()?;
                _arg2_set = true;
                reader.context().pop();
            }
            err => return Err(DecodeError::UnknownFieldName(err.to_string())),
        }
    }
    if !_arg1_set {
        return Err(DecodeError::MissingField("arg1: UInt32.".to_string()));
    }
    if !_arg2_set {
        return Err(DecodeError::MissingField("arg2: Boolean.".to_string()));
    }

    Ok(ArgsFunction1 {
        arg1: _arg1,
        arg2: _arg2,
    })
}

pub fn serialize_function1_args(args: &ArgsFunction1) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) module-type: function1 Args".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_function1_args(args, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_function1_args<W: Write>(args: &ArgsFunction1, writer: &mut W) -> Result<(), EncodeError> {
    writer.write_map_length(&2)?;
    writer.context().push("arg1", "u32", "writing property");
    writer.write_string("arg1")?;
    writer.write_u32(&args.arg1)?;
    writer.context().pop();
    writer.context().push("arg2", "bool", "writing property");
    writer.write_string("arg2")?;
    writer.write_bool(&args.arg2)?;
    writer.context().pop();
    Ok(())
}

pub fn serialize_function1_result(result: &String) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) module-type: function1 Result".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_function1_result(result, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_function1_result<W: Write>(result: &String, writer: &mut W) -> Result<(), EncodeError> {
    writer.context().push("function1", "String", "writing result");
    writer.write_string(result)?;
    writer.context().pop();
    Ok(())
}

pub fn deserialize_function1_result(result: &[u8]) -> Result<String, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing module-type: function1 Result".to_string();
    let mut reader = ReadDecoder::new(result, context);

    reader.context().push("function1", "String", "reading function output");
    let res = reader.read_string()?;
    reader.context().pop();
    Ok(res)
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ArgsFunction2 {
    pub arg1: Option<i32>,
    pub arg2: Option<Vec<u8>>,
}

pub fn deserialize_function2_args(args: &[u8]) -> Result<ArgsFunction2, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing module-type: function2 Args".to_string();

    let mut reader = ReadDecoder::new(args, context);
    let mut num_of_fields = reader.read_map_length()?;

    let mut _arg1: Option<i32> = None;
    let mut _arg2: Option<Vec<u8>> = None;

    while num_of_fields > 0 {
        num_of_fields -= 1;
        let field = reader.read_string()?;

        match field.as_str() {
            "arg1" => {
                reader.context().push(&field, "Option<i32>", "type found, reading argument");
                _arg1 = reader.read_optional_i32()?;
                reader.context().pop();
            }
            "arg2" => {
                reader.context().push(&field, "Option<Vec<u8>>", "type found, reading argument");
                _arg2 = reader.read_optional_bytes()?;
                reader.context().pop();
            }
            err => return Err(DecodeError::UnknownFieldName(err.to_string())),
        }
    }

    Ok(ArgsFunction2 {
        arg1: _arg1,
        arg2: _arg2,
    })
}

pub fn serialize_function2_args(args: &ArgsFunction2) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) module-type: function2 Args".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_function2_args(args, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_function2_args<W: Write>(args: &ArgsFunction2, writer: &mut W) -> Result<(), EncodeError> {
    writer.write_map_length(&2)?;
    writer.context().push("arg1", "Option<i32>", "writing property");
    writer.write_string("arg1")?;
    writer.write_optional_i32(&args.arg1)?;
    writer.context().pop();
    writer.context().push("arg2", "Option<Vec<u8>>", "writing property");
    writer.write_string("arg2")?;
    writer.write_optional_bytes(&args.arg2)?;
    writer.context().pop();
    Ok(())
}

pub fn serialize_function2_result(result: &Option<String>) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) module-type: function2 Result".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_function2_result(result, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_function2_result<W: Write>(result: &Option<String>, writer: &mut W) -> Result<(), EncodeError> {
    writer.context().push("function2", "Option<String>", "writing result");
    writer.write_optional_string(result)?;
    writer.context().pop();
    Ok(())
}

pub fn deserialize_function2_result(result: &[u8]) -> Result<Option<String>, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing module-type: function2 Result".to_string();
    let mut reader = ReadDecoder::new(result, context);

    reader.context().push("function2", "Option<String>", "reading function output");
    let res = reader.read_optional_string()?;
    reader.context().pop();
    Ok(res)
}

File: implementations/wrap-assemblyscript/src/__tests__/cases/001-module-functions/output/Module/serialization.ts
import {
  Read,
  ReadDecoder,
  Write,
  WriteSizer,
  WriteEncoder,
  Box,
  BigInt,
  BigNumber,
  JSON,
  Context
} from "@polywrap/wasm-as";
import * as Types from "..";

export class Args_function1 {
  arg1: u32;
  arg2: bool;
}

export function deserializefunction1Args(argsBuf: ArrayBuffer): Args_function1 {
  const context: Context = new Context("Deserializing module-type: function1 Args");
  const reader = new ReadDecoder(argsBuf, context);
  let numFields = reader.readMapLength();

  let _arg1: u32 = 0;
  let _arg1Set: bool = false;
  let _arg2: bool = false;
  let _arg2Set: bool = false;

  while (numFields > 0) {
    numFields--;
    const field = reader.readString();

    reader.context().push(field, "unknown", "searching for property type");
    if (field == "arg1") {
      reader.context().push(field, "u32", "type found, reading property");
      _arg1 = reader.readUInt32();
      _arg1Set = true;
      reader.context().pop();
    }
    else if (field == "arg2") {
      reader.context().push(field, "bool", "type found, reading property");
      _arg2 = reader.readBool();
      _arg2Set = true;
      reader.context().pop();
    }
    reader.context().pop();
  }

  if (!_arg1Set) {
    throw new Error(reader.context().printWithContext("Missing required argument: 'arg1: UInt32'"));
  }
  if (!_arg2Set) {
    throw new Error(reader.context().printWithContext("Missing required argument: 'arg2: Boolean'"));
  }

  return {
    arg1: _arg1,
    arg2: _arg2
  };
}

export function serializefunction1Args(args: Args_function1): ArrayBuffer {
  const sizerContext: Context = new Context("Serializing (sizing) module-type: function1 Args");
  const sizer = new WriteSizer(sizerContext);
  writefunction1Args(sizer, args);
  const buffer = new ArrayBuffer(sizer.length);
  const encoderContext: Context = new Context("Serializing (encoding) module-type: function1 Args");
  const encoder = new WriteEncoder(buffer, sizer, encoderContext);
  writefunction1Args(encoder, args);
  return buffer;
}

export function writefunction1Args(
  writer: Write,
  args: Args_function1
): void {
  writer.writeMapLength(2);
  writer.context().push("arg1", "u32", "writing property");
  writer.writeString("arg1");
  writer.writeUInt32(args.arg1);
  writer.context().pop();
  writer.context().push("arg2", "bool", "writing property");
  writer.writeString("arg2");
  writer.writeBool(args.arg2);
  writer.context().pop();
}

export function serializefunction1Result(result: string): ArrayBuffer {
  const sizerContext: Context = new Context("Serializing (sizing) module-type: function1 Result");
  const sizer = new WriteSizer(sizerContext);
  writefunction1Result(sizer, result);
  const buffer = new ArrayBuffer(sizer.length);
  const encoderContext: Context = new Context("Serializing (encoding) module-type: function1 Result");
  const encoder = new WriteEncoder(buffer, sizer, encoderContext);
  writefunction1Result(encoder, result);
  return buffer;
}

export function writefunction1Result(writer: Write, result: string): void {
  writer.context().push("function1", "string", "writing property");
  writer.writeString(result);
  writer.context().pop();
}

export function deserializefunction1Result(buffer: ArrayBuffer): string {
  const context: Context = new Context("Deserializing module-type: function1 Result");
  const reader = new ReadDecoder(buffer, context);

  reader.context().push("function1", "string", "reading function output");
  const res: string = reader.readString();
  reader.context().pop();

  return res;
}

export class Args_function2 {
  arg1: Box<i32> | null;
  arg2: ArrayBuffer | null;
}

export function deserializefunction2Args(argsBuf: ArrayBuffer): Args_function2 {
  const context: Context = new Context("Deserializing module-type: function2 Args");
  const reader = new ReadDecoder(argsBuf, context);
  let numFields = reader.readMapLength();

  let _arg1: Box<i32> | null = null;
  let _arg2: ArrayBuffer | null = null;

  while (numFields > 0) {
    numFields--;
    const field = reader.readString();

    reader.context().push(field, "unknown", "searching for property type");
    if (field == "arg1") {
      reader.context().push(field, "Box<i32> | null", "type found, reading property");
      _arg1 = reader.readOptionalInt32();
      reader.context().pop();
    }
    else if (field == "arg2") {
      reader.context().push(field, "ArrayBuffer | null", "type found, reading property");
      _arg2 = reader.readOptionalBytes();
      reader.context().pop();
    }
    reader.context().pop();
  }

  return {
    arg1: _arg1,
    arg2: _arg2
  };
}

export function serializefunction2Args(args: Args_function2): ArrayBuffer {
  const sizerContext: Context = new Context("Serializing (sizing) module-type: function2 Args");
  const sizer = new WriteSizer(sizerContext);
  writefunction2Args(sizer, args);
  const buffer = new ArrayBuffer(sizer.length);
  const encoderContext: Context = new Context("Serializing (encoding) module-type: function2 Args");
  const encoder = new WriteEncoder(buffer, sizer, encoderContext);
  writefunction2Args(encoder, args);
  return buffer;
}

export function writefunction2Args(
  writer: Write,
  args: Args_function2
): void {
  writer.writeMapLength(2);
  writer.context().push("arg1", "Box<i32> | null", "writing property");
  writer.writeString("arg1");
  writer.writeOptionalInt32(args.arg1);
  writer.context().pop();
  writer.context().push("arg2", "ArrayBuffer | null", "writing property");
  writer.writeString("arg2");
  writer.writeOptionalBytes(args.arg2);
  writer.context().pop();
}

export function serializefunction2Result(result: string | null): ArrayBuffer {
  const sizerContext: Context = new Context("Serializing (sizing) module-type: function2 Result");
  const sizer = new WriteSizer(sizerContext);
  writefunction2Result(sizer, result);
  const buffer = new ArrayBuffer(sizer.length);
  const encoderContext: Context = new Context("Serializing (encoding) module-type: function2 Result");
  const encoder = new WriteEncoder(buffer, sizer, encoderContext);
  writefunction2Result(encoder, result);
  return buffer;
}

export function writefunction2Result(writer: Write, result: string | null): void {
  writer.context().push("function2", "string | null", "writing property");
  writer.writeOptionalString(result);
  writer.context().pop();
}

export function deserializefunction2Result(buffer: ArrayBuffer): string | null {
  const context: Context = new Context("Deserializing module-type: function2 Result");
  const reader = new ReadDecoder(buffer, context);

  reader.context().push("function2", "string | null", "reading function output");
  const res: string | null = reader.readOptionalString();
  reader.context().pop();

  return res;
}
